### ISSUE TITLE
Autocomplete for __getattr__ by Literal values

### ISSUE THREAD

**Issue by tkukushkin** [microsoft/pylance-release#4660]
For example I have following class:

```python
from typing import Any, Literal, overload


class A:

    @overload
    def __getattr__(self, item: Literal['foo', 'bar']) -> int:
        pass

    @overload
    def __getattr__(self, item: Literal['baz']) -> str:
        pass

    @overload
    def __getattr__(self, item: str) -> int | str:
        pass

    def __getattr__(self, item: str) -> Any:
        return None
```

I would be really nice if pylance could autocomplete attributes by names from Literal.

So for example I write following code:

```python
a = A()
a.ba<ctrl+space>
```
And I would like to see `bar` in autocomplete suggestions.

Current behaviour:
<img width="538" alt="image" src="https://github.com/microsoft/pylance-release/assets/1482516/5b90b08d-0722-452d-b124-edd2c27cdcc7">

**Comment by escummy** [comment]
As far as I know, Pylance might not fully support autocompletion for attributes defined using Literal in overloaded __getattr__ methods. 

> However, we can try to address this by providing additional type information that Pylance can recognize.

To achieve autocompletion for __getattr__ with Literal values, we can define a custom type hint that represents the available attribute names. Let's create a LiteralAttributes type hint and use it in the __getattr__ overloads:

```python
from typing import Any, Literal, overload, Union

# Define a custom LiteralAttributes type hint
LiteralAttributes = Literal['foo', 'bar', 'baz']

class A:

    @overload
    def __getattr__(self, item: LiteralAttributes) -> int:
        pass

    @overload
    def __getattr__(self, item: Literal['baz']) -> str:
        pass

    @overload
    def __getattr__(self, item: str) -> Union[int, str]:
        pass

    def __getattr__(self, item: str) -> Any:
        return None

# Create an instance of A
a = A()
```

Now, when you type 'a.ba' and press Ctrl+Space, you should see 'bar' in the autocompletion suggestions.
By defining a custom LiteralAttributes type hint, we provide explicit information to Pylance about the possible attribute names for autocompletion. 

> This should enable Pylance to show the appropriate suggestions when using overloaded __getattr__ methods with Literal.

**Comment by rchiodo** [comment]
@escummy I'm not sure why you think this would work. I believe the reason this isn't possible is because `__getattr__` is not used internally to determine possible members. Are you just posting what chat GPT says the solution to these errors are? Your posts seem to be hallucinating features in Pylance. I thought that might be because you're not actually validating the truth of your statements and just outputting what chat GPT is saying. 

You can try the code you posted above with the latest version of Pylance. `a.ba` does not complete to anything:

![image](https://github.com/microsoft/pylance-release/assets/19672699/fec131ad-734f-4a5d-82e8-723a465a22c9)

**Comment by escummy** [comment]
That`s not he case @rchiodo but if you prefer I can delete the answer for this issue.

**Comment by rchiodo** [comment]
No just please validate solutions before you post them. You mentioned a non existent setting in this post here:
https://github.com/microsoft/pylance-release/issues/4627#issuecomment-1662515904

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.8.51, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023851-30-August-2023-PreRelease)
